text/0000-svelte-config-util.md

```diff
@@ -0,0 +1,165 @@
+- Start Date: (fill me in with today's date, 2021-09-28)
+- RFC PR: (leave this empty)
+- Svelte Issue: (leave this empty)
+
+# Svelte config utility
+
+## Summary
+
+Introduce a new utility package `@sveltejs/config` to offer tools in the svelte ecosystem a common way
+to read and validate svelte config files and provide config features to users in a consistent manner.
+
+## Motivation
+
+Currently, tools implement config reading separately (or not at all),
+leading to extra maintenance and more work for new tools who need to read it.
+
+Reducing this overhead and offering a place to enforce consistency
+and implement new features is the main motivation of this RFC,
+resulting in a better dx for both users writing svelte config files
+and tool authors consuming them via a simple common api.
+
+### Features
+
+#### for tools
+* async `loadConfig` function
+* defined namespacing (top-level keys owned by a tool)
+* validation utilities
+
+#### for users
+* `defineConfig` helper for intellisense like in [vite config](https://vitejs.dev/config/#config-intellisense)
+* `async` config support like in [vite config](https://vitejs.dev/config/#async-config)
+* allow `svelte.config.ts`
+
+
+## Detailed design
+
+### Technical Background
+
+### Prior Art
+* [postcss-load-config](https://github.com/postcss/postcss-load-config)
+* [sveltekit](https://github.com/sveltejs/kit/tree/master/packages/kit/src/core/config)
+* [vite-plugin-svelte](https://github.com/sveltejs/vite-plugin-svelte/blob/main/packages/vite-plugin-svelte/src/utils/load-svelte-config.ts)
+
+### Implementation
+
+The implementation is going to be fully compatible to existing usage of svelte.config.js (cjs/mjs) files.
+
+Tools can opt in by adopting `@sveltejs/config`
+
+New config features for users could potentially introduce changes that are not backwards compatible
+to tools reading svelte config without `@sveltejs/config`.
+To reduce friction, v1.0 of @sveltejs/config could be limited to features that are explicitly backwards compatible
+and features that would break will only be released after our tools adopted it
+
+#### New Features for tool authors
+
+##### async `loadConfig` function
+```js
+import { loadConfig } from '@sveltejs/config';
+//...
+const config = await load_config();
+```
+
+loadConfig accepts an optional `options` argument with the following options:
+```js
+options = {
+    fromDir: '/some/path', // process.cwd by default, tools can pass what they need, eg vite root
+    inlineConfig: {/*...*/}, // svelte config object built from inline options eg cli args passed to the tool. merged in
+    namespace: 'kit', // namespace for the tool loading the config, useful for advanced features
```

Comment:
Could you elaborate this one?

Reply:
I mean this one

```diff
+      namespace: 'kit', // namespace for the tool loading the config, useful for advanced features
```

Reply:
if sveltekit calls `@sveltejs/config`  to load the svelte config, it can identify itself which enables sveltejs/config to implement some kit specific features or just have better logging.